### PR TITLE
Feat statistics

### PR DESCRIPTION
**Pull request by thommann**

### CODE REVIEW

File: Simulator/Simulator.py [hunk 1]
```diff
@@ -40,31 +44,32 @@ def generate_new_agents(self) -> Dict[int, "Agent"]:
...
+        Returns `True` until the simulation is done, then it returns `False`.
+        :return:
+        """
         if self.time_step > self.environment.dimension.t:
             return False
 
         new_agents: Dict[int, "Agent"] = self.generate_new_agents()
 
-        if len(new_agents) > 0 or self.allocated_last_tick:
+        if len(new_agents) > 0 or self.mechanism.allocator.wants_to_reallocate():
```
Comment (johannschwabe) [inline Simulator/Simulator.py]: pass environment

Comment (johannschwabe) [reply Simulator/Simulator.py]: and tick

File: Simulator/Statistics/Statistics.py [hunk 2]
```diff
@@ -12,154 +15,327 @@
...
+    L1_DISTANCE = "l1_distance"
+    L2_DISTANCE = "l2_distance"
+    L1_GROUND_DISTANCE = "l1_ground_distance"
+    L2_GROUND_DISTANCE = "l2_ground_distance"
+    HEIGHT_DIFFERENCE = "height_difference"
+    TIME_DIFFERENCE = "time_difference"
+    ASCENT = "ascent"
+    DESCENT = "descent"
+    L1_DISTANCE_TRAVELED = "l1_distance_traveled"
+    L2_DISTANCE_TRAVELED = "l2_distance_traveled"
```
Comment (johannschwabe) [inline Simulator/Statistics/Statistics.py]: same as L1

File: API/API.py [hunk 3]
```diff
@@ -33,6 +34,21 @@
...
+    statistics_duration = statistics_end_time - statistics_start_time
+    statistics_compute_time = statistics_duration
+    simulation_compute_time = simulation_duration
+    return {"config": config.dict(),
+            "simulation": simulation_json,
+            "statistics": statistics,
+            "statistics_compute_time": statistics_compute_time,
+            "simulation_compute_time": simulation_compute_time}
+
+
```
Comment (johannschwabe) [inline API/API.py]: assure compatibility with CLI

File: Development/playground.py [hunk 4]
```diff
@@ -39,27 +41,34 @@ def setup_map():
                                   math.floor(size[1]),
                                   time_steps)
     print(dimensions)
-    return EnvironmentGen(map_dimensions, [MapTile([15, 17161, 11475], area)]).generate()
+    return EnvironmentGen(map_dimensions, [MapTile([15, 17161, 11475], area)], 50, 10).generate()
```
Comment (johannschwabe) [inline Development/playground.py]: needs new arguments

File: Simulator/AStar/AStar.py [hunk 5]
```diff
@@ -154,16 +159,13 @@ def is_valid_for_allocation(self, position, agent):
                 return True, colliding_agents
             return False, None
 
-        agent_hashes = self.environment.intersect(position, agent.near_radius, agent.speed)
-        my_bid = self.bid_tracker.get_last_bid_for_tick(self.tick, agent, self.environment)
-
-        if my_bid is None:
-            return False, None
+        colliding_agents = self.environment.intersect_path_coordinate(position, agent)
```
Comment (johannschwabe) [inline Simulator/AStar/AStar.py]: naming chaos: var name colliding_agents is already taken

File: Simulator/Environment/Environment.py [hunk 6]
```diff
@@ -19,50 +20,46 @@ class Environment:
...
-                 _tree: Optional[Index] = None,
-                 _blocker_tree: Optional[Index] = None):
+                 min_height: int = 0,
+                 allocation_period: Optional[int] = None):
 
         self.dimension: "Coordinate4D" = dimension
         self.min_height = min_height
+
+        if allocation_period is None:
+            allocation_period = dimension.t
```
Comment (johannschwabe) [inline Simulator/Environment/Environment.py]: doesn't make sense to allocate agents in the last tick

File: Simulator/Environment/Environment.py [hunk 7]
```diff
@@ -175,18 +172,17 @@ def allocate_segments_for_agents(self,
...
             elif isinstance(agent, PathAgent):
-                segments: List["PathSegment"] = allocation.segments
-                self.register_agent(agent, time_step)
                 self.allocate_path_for_agent(agent, segments)
+
             else:
-                raise Exception(f"Unknown Agent: {agent}")
+                raise Exception(f"Invalid Agent: {agent}")
 
     def register_agent(self, agent: "Agent", time_step: int):
```
Comment (johannschwabe) [inline Simulator/Environment/Environment.py]: no longer just registering agents, new name needed

File: Simulator/IO/Statistics.py [hunk 8]
```diff
@@ -0,0 +1,680 @@
...
+        heights = []
+        times = []
+        heights_above_ground = []
+
+        tree = Statistics._setup_rtree()
+
+        for space_segment in spaces:
+            intersections = tree.intersection(space_segment.tree_rep(), objects="raw")
+            for intersecting_space_segment in intersections:
+                assert isinstance(intersecting_space_segment, SpaceSegment)
```
Comment (johannschwabe) [inline Simulator/IO/Statistics.py:345]: Should selfcollisions be counted?

File: Simulator/IO/Statistics.py [hunk 9]
```diff
@@ -0,0 +1,680 @@
...
+        violations: Dict[str, List["Coordinate4D"]] = {}
+        total_violations: int = 0
+
+        for coordinate in path_segment.coordinates:
+            intersecting_agents = self.simulation.environment.intersect_path_coordinate(coordinate,
+                                                                                        path_agent,
+                                                                                        include_speed=False)
+            for intersecting_agent in intersecting_agents:
+                if intersecting_agent != path_agent:
+                    if isinstance(intersecting_agent, PathAgent):
```
Comment (johannschwabe) [inline Simulator/IO/Statistics.py]: self intersections

File: Visualizer/src/API/api.js [hunk 10]
```diff
@@ -94,31 +96,119 @@ export async function getPaymentRulesSupportedByAllocator(allocator) {
   }
 }
 
+async function openDB() {
```
Comment (johannschwabe) [inline Visualizer/src/API/api.js:99]: maybe move db helper to seperate file?

**Comment by joelbarmettlerUZH** [comment]
lgtm

**Review by johannschwabe — CHANGES_REQUESTED** [review]
some smoll considerations

File: CLI.py [hunk 11]
```diff
@@ -244,7 +246,8 @@ def all_bidding_strategies_and_value_functions_str() -> str:
                          'air at the same time. However, keep in mind that large playfields with fine granularity '
                          'require the agents to fly for longer. Agents that can not reach their destination within '
                          'your specified timesteps will not be allocated.')
-parser.add_argument('--allocation-period', dest="allocationPeriod", type=int, choices=range(300, 4000), metavar="[300, 1000]",
+parser.add_argument('--allocation-period', dest="allocationPeriod", type=int, choices=range(300, 4000),
+                    metavar="[300, 1000]",
```
Comment (joelbarmettlerUZH) [inline CLI.py]:
```suggestion
parser.add_argument('--allocation-period', dest="allocationPeriod", type=int, choices=range(300, 4000),
                    metavar="[300, 4000]",
```

File: Simulator/IO/Statistics.py [hunk 12]
```diff
@@ -0,0 +1,681 @@
...
+        """
+        Calculate the value for the allocations of all agents of an owner summed up.
+        :param owner:
+        :return:
+        """
+        values = [self.get_value_for_agent(agent) for agent in owner.agents]
+        return self._get_value_statistics(values)
+
+    @staticmethod
+    def path_segment_statistics(path_segment: "PathSegment"):
```
Comment (joelbarmettlerUZH) [inline Simulator/IO/Statistics.py]: indicate return type "PathStatistics"

File: Simulator/IO/Statistics.py [hunk 13]
```diff
@@ -0,0 +1,681 @@
...
+                              ascent,
+                              descent,
+                              distance_traveled,
+                              ground_distance_traveled,
+                              mean_height,
+                              median_height,
+                              heights)
+
+    @staticmethod
+    def space_segment_statistics(space_segment: "SpaceSegment"):
```
Comment (joelbarmettlerUZH) [inline Simulator/IO/Statistics.py]: indicate return type "SpaceSegmentStatistics"

**Review by joelbarmettlerUZH — APPROVED** [review]
lgtm, minor stuff

**Review by johannschwabe — APPROVED** [review]

**Review by johannschwabe — APPROVED** [review]

**Pull request merged**